### ISSUE TITLE
Microprofile with JPA

### ISSUE THREAD

**Issue by jShiwaniGupta** [payara/Payara#1283]
# Description #
----------
Trying microprofile payara version with jpa eclipselink but getting the following error on JPA operation.

**JPA Dependencies :**
````
<dependency>
    <groupId>org.eclipse.persistence</groupId>
    <artifactId>eclipselink</artifactId>
    <version>2.6.4</version>
</dependency>

<dependency>
  <groupId>javax.persistence</groupId>
  <artifactId>persistence-api</artifactId>
  <version>1.0.2</version>
</dependency>
````
**Payara profile :**
````
    <profile>
            <id>micro</id>
            <build>
                <plugins>
                    <plugin>
                        <groupId>org.codehaus.mojo</groupId>
                        <artifactId>exec-maven-plugin</artifactId>
                        <version>1.5.0</version>
                        <dependencies>
                            <dependency>
                                <groupId>fish.payara.extras</groupId>
                                <artifactId>payara-microprofile</artifactId>
                                <version>1.0</version>
                            </dependency>
                        </dependencies>
                        <executions>
                            <execution>
                                <id>payara-uber-jar</id>
                                <phase>package</phase>
                                <goals>
                                    <goal>java</goal>
                                </goals>
                                <configuration>
                                    <mainClass>fish.payara.micro.PayaraMicro</mainClass>
                                    <arguments>
                                        <argument>--deploy</argument>
                                        <argument>${basedir}/target/${project.build.finalName}.war</argument>
                                        <argument>--outputUberJar</argument>                                                  
                                        <argument>${basedir}/target/${project.build.finalName}.jar</argument>
                                    </arguments>
                                    <includeProjectDependencies>false</includeProjectDependencies>
                                    <includePluginDependencies>true</includePluginDependencies>
                                    <executableDependency>
                                        <groupId>fish.payara.extras</groupId>
                                        <artifactId>payara-microprofile</artifactId>
                                    </executableDependency>
                                </configuration>
                            </execution>
                        </executions>
                    </plugin>
                </plugins>
            </build>
        </profile>
   ````
**persistence.xml :** (src\main\resources\META-INF\persistence.xml)
````
<?xml version="1.0" encoding="UTF-8"?>
<persistence version="2.1" xmlns="http://xmlns.jcp.org/xml/ns/persistence" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/persistence http://xmlns.jcp.org/xml/ns/persistence/persistence_2_1.xsd">
  <persistence-unit name="DEFAULT_PU" transaction-type="JTA">
    <provider>org.eclipse.persistence.jpa.PersistenceProvider</provider>
    <jta-data-source>jdbc/sample</jta-data-source>
    <class>domain.Entity_1</class>
    <class>domain.User</class>
    <class>domain.Authority</class>
    <exclude-unlisted-classes>false</exclude-unlisted-classes>
    <properties>
      <property name="javax.persistence.schema-generation.database.action" value="drop-and-create"/>
      <property name="javax.persistence.sql-load-script-source" value="META-INF/sql/insert.sql"/>
    </properties>
  </persistence-unit>
</persistence>
````

## Current Outcome
````
 StandardWrapperValve[controller.ApplicationConfig]: Servlet.service() for servlet controller.ApplicationConfig threw exception
java.lang.IllegalStateException: Unable to retrieve EntityManagerFactory for unitName DEFAULT_PU
        at com.sun.enterprise.container.common.impl.EntityManagerWrapper.init(EntityManagerWrapper.java:138)
        at com.sun.enterprise.container.common.impl.EntityManagerWrapper._getDelegate(EntityManagerWrapper.java:171)
        at com.sun.enterprise.container.common.impl.EntityManagerWrapper.createNamedQuery(EntityManagerWrapper.java:536)
        at service.facade.AbstractFacade.findSingleByNamedQuery(AbstractFacade.ava:85)
        at service.facade.UserFacade.findOneWithAuthoritiesByLogin(UserFacade.java:53)
        at app.service.UserService.authenticate(UserService.java:180)
        at controller.UserJWTController.login(UserJWTController.java:62)
        at controller.UserJWTController$Proxy$_$$_WeldSubclass.login$$super

````

## Environment 

- **Payara Microprofile Version**: 1.0
- **JDK Version**: 8 
- **Operating System**: Windows

Is any Payara[JPA + Microprofile] example available ?

**Comment by mikecroft** [comment]
Is there a reason why you can't use Payara Micro instead? Microprofile just has 3 APIs in it right now, but Payara Micro has the same APIs as the web profile plus JCache, JBatch and Concurrency.

Payara Micro is where we started with this, so we released Microprofile as just a subset of that. That means that all the same command line options should work in both and both can be used in the same way.

**Comment by jGauravGupta** [comment]
Hi @mikecroft ,

The reason is we are trying to provide the support for microprofile in JPA Modeler so that other developers can experiment with generated running Payara Microprofile without any hassle. 

I was looking at the source(Payara\appserver\extras\payara-micro\payara-micro-microprofile-distribution) the only difference is in config and dependencies. 

So the problem is from Payara Microprofile side or in the above code ? Can you give some hints .

**Comment by smillidge** [comment]
Payara Microprofile has only been created as part of the microprofile.io initiative to have the 3 microprofile apis and that does not include JPA or the ability to create datasources or anything like that as it only includes CDI, JAX-RS and JSON-P. 

If you want to test JPA on Payara Micro you need to use normal Payara Micro.

**Comment by jGauravGupta** [comment]
Thanks @smillidge for the response, So the problem is datasource handling.

I understand that microprofile have inbuilt support for CDI, JAX-RS and JSON-P but the following microprofile implementation provides integration for persistence providers(eclipselink, hibernate etc), e.g :
- org.wildfly.swarm :: jpa-eclipselink
- ws.ament.hammock :: jpa-eclipselink

So it would be great to have pluggable JPA implementation &  datasource handling support in **Payara Microprofile** because for a hibernate user it bundled eclipselink jpa in Payara Micro will not be useful.
- MicroProfile : 32.63MB
- Micro : 57.53MB

Ref : https://issues.jboss.org/browse/SWARM-481

**Comment by OndrejM** [comment]
MicroProfile currently does not officially support JPA at all. The fractions provided by WF Swarm and Hammock are vendor specific. We are working on providing a similar feature into Payara Microprofile too, but it's not there yet. If you want to use JPA, you either have to use a RESOURCE_LOCAL transaction type, or also bundle a JTA provider, because it is not included in the MicroProfile too.

**Comment by jGauravGupta** [comment]
Hi @OndrejM , 
I was looking at the content of microprofile jar, it have jpa specification class inside it, May be some other dependency depends on it. So that JPA spec classes are still bundled inside microprofile jar. Not only the JPA but also observing some other unwanted classes/spec that should not be the part of micro-profile.
Correct me if i am wrong :)

**Comment by OndrejM** [comment]
You are correct, @jGauravGupta, I didn't know about that. This is not intended, but it is not easy to get rid of all unnecessary dependencies. We are seeking way how to make the MicroProfile version even smaller.

But the message I wanted to send is that, with the MicroProfile distribution, you shouldn't rely on anything but what is defined by the MicroProfile. This is currently only 3 specs: CDI, JAX-RS and JSON-Processing. Everything else is unsupported and undocumented, and may be removed in the future. The applications should either rely on those 3 specs or on plain Java SE.

If you have issues running an application that includes hibernate or eclipselink with the MicroProfile distribution, raise a concrete issue. 

Or try using Payara Micro, which is also a MicroProfile implementation, but includes much more (in respect to MicroProfile you may call it vendor specific, even if it is Java EE API ;) ) If you have issues using Payara Micro with a MicroPofile app, we'll be happy to help you.

**Comment by jShiwaniGupta** [comment]
Has this issue been resolved in new version 4.1.1.171 ?

**Comment by smillidge** [comment]
@jShiwaniGupta this is not really an issue to fix. Payara Micro ships with JPA, microprofile doesn't and you should use the apis in microprofile only. We are introducing a composable container that allows JPA to be added on top of microprofile in a supported way by adding jars into the package.
